### DISCUSSION TITLE
<unknown> java: org.libpag.PAGImageView.c(SourceFile:33) org.libpag.PAGImageView.onVisibilityAggregated(SourceFile:4)

### DISCUSSION THREAD

**Discussion by chaychan** [Tencent/libpag#2571] [Q&A]
SDK版本：4.3.62
安卓版本：11
品牌：OPPO 
型号：CPH2349

麻烦帮忙定位下问题，虽然是java的崩溃，PagImageView中代码都是混淆的，看不出问题的所在，麻烦知道的技术大大赐教
以下是崩溃日志：

`pc 0000000000078974    /apex/com.android.runtime/lib64/bionic/libc.so (abort+180) [arm64-v8a::a6e89496b61a77b8a97118405d4e247e]
#01    pc 00000000000dc9d4    /apex/com.android.runtime/lib64/bionic/libc.so (_ZL15__fortify_fatalPKcz+124) [arm64-v8a::a6e89496b61a77b8a97118405d4e247e]
#02    pc 00000000000dbebc    /apex/com.android.runtime/lib64/bionic/libc.so (_ZL25HandleUsingDestroyedMutexP15pthread_mutex_tPKc+52) [arm64-v8a::a6e89496b61a77b8a97118405d4e247e]
#03    pc 00000000000dbd14    /apex/com.android.runtime/lib64/bionic/libc.so (pthread_mutex_lock+172) [arm64-v8a::a6e89496b61a77b8a97118405d4e247e]
#04    pc 000000014b37229c    /data/app/~~xR_MV96FsSRTYgXAkqisFQ==/com.voice-Vz-SYQWJKVrXI6nIta-HcA==/split_config.arm64_v8a.apk
#05    pc 000000014b22345c    /data/app/~~xR_MV96FsSRTYgXAkqisFQ==/com.voice-Vz-SYQWJKVrXI6nIta-HcA==/split_config.arm64_v8a.apk (_ZNK3pag8PAGLayer8durationEv+68)
#06    pc 000000014b24cedc    /data/app/~~xR_MV96FsSRTYgXAkqisFQ==/com.voice-Vz-SYQWJKVrXI6nIta-HcA==/split_config.arm64_v8a.apk (Java_org_libpag_PAGLayer_duration+68)
#07    pc 000000009bf7658c    <unknown>
java:
org.libpag.PAGImageView.c(SourceFile:33)
org.libpag.PAGImageView.onVisibilityAggregated(SourceFile:4)
android.view.View.dispatchVisibilityAggregated(View.java:14967)
android.view.ViewGroup.dispatchVisibilityAggregated(ViewGroup.java:1639)
android.view.ViewGroup.dispatchVisibilityAggregated(ViewGroup.java:1639)
android.view.ViewGroup.dispatchVisibilityAggregated(ViewGroup.java:1639)
android.view.ViewGroup.attachViewToParent(ViewGroup.java:6058)
androidx.recyclerview.widget.RecyclerView.access$000(Unknown Source:0)
androidx.recyclerview.widget.RecyclerView$5.attachViewToParent(SourceFile:47)
androidx.recyclerview.widget.ChildHelper.attachViewToParent(SourceFile:26)
androidx.recyclerview.widget.RecyclerView$LayoutManager.addViewInt(SourceFile:155)
androidx.recyclerview.widget.RecyclerView$LayoutManager.addView(SourceFile:2)
androidx.recyclerview.widget.RecyclerView$LayoutManager.addView(SourceFile:1)
androidx.recyclerview.widget.GridLayoutManager.layoutChunk(SourceFile:157)
androidx.recyclerview.widget.LinearLayoutManager.fill(SourceFile:39)
androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren(SourceFile:357)
androidx.recyclerview.widget.GridLayoutManager.onLayoutChildren(SourceFile:10)
androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep2(SourceFile:67)
androidx.recyclerview.widget.RecyclerView.onMeasure(SourceFile:68)
android.view.View.measure(View.java:26027)
androidx.constraintlayout.widget.ConstraintLayout$c.b(SourceFile:548)
androidx.constraintlayout.core.widgets.d.Y(SourceFile:265)
androidx.constraintlayout.core.widgets.analyzer.g.b(SourceFile:33)
androidx.constraintlayout.core.widgets.analyzer.g.d(SourceFile:104)
androidx.constraintlayout.core.widgets.analyzer.g.b(SourceFile:188)
androidx.constraintlayout.core.widgets.d.V(SourceFile:254)
androidx.constraintlayout.core.widgets.analyzer.b.b(SourceFile:34)
androidx.constraintlayout.widget.ConstraintLayout.resolveSystem(SourceFile:701)
androidx.constraintlayout.widget.ConstraintLayout.onMeasure(SourceFile:65)
android.view.View.measure(View.java:26027)
androidx.viewpager.widget.ViewPager.onMeasure(SourceFile:238)
android.view.View.measure(View.java:26027)
androidx.constraintlayout.widget.ConstraintLayout$c.b(SourceFile:548)
androidx.constraintlayout.core.widgets.analyzer.b.a(SourceFile:93)
androidx.constraintlayout.widget.ConstraintLayout.resolveSystem(SourceFile:675)
androidx.constraintlayout.widget.ConstraintLayout.onMeasure(SourceFile:65)
android.view.View.measure(View.java:26027)
android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7134)
android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
androidx.appcompat.widget.ContentFrameLayout.onMeasure(SourceFile:160)
android.view.View.measure(View.java:26027)
android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7134)
android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1552)
android.widget.LinearLayout.measureVertical(LinearLayout.java:842)
android.widget.LinearLayout.onMeasure(LinearLayout.java:721)
android.view.View.measure(View.java:26027)
android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7134)
android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
android.view.View.measure(View.java:26027)
android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7134)
android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1552)
android.widget.LinearLayout.measureVertical(LinearLayout.java:842)
android.widget.LinearLayout.onMeasure(LinearLayout.java:721)
android.view.View.measure(View.java:26027)
android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7134)
android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
com.android.internal.policy.DecorView.onMeasure(DecorView.java:835)
android.view.View.measure(View.java:26027)
android.view.ViewRootImpl.performMeasure(ViewRootImpl.java:3915)
android.view.ViewRootImpl.measureHierarchy(ViewRootImpl.java:2634)
android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2897)
android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:2354)
android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:9097)
android.view.Choreographer$CallbackRecord.run(Choreographer.java:1458)
android.view.Choreographer.doCallbacks(Choreographer.java:1256)
android.view.Choreographer.doFrame(Choreographer.java:953)
android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1443)
android.os.Handler.handleCallback(Handler.java:938)
android.os.Handler.dispatchMessage(Handler.java:99)
android.os.Looper.loop(Looper.java:264)
android.app.ActivityThread.main(ActivityThread.java:8315)
java.lang.reflect.Method.invoke(Native Method)
com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:632)
com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1049)`

**Comment by L-Evan** [comment]
也遇到了

**Reply by L-Evan** [reply]
03-24 18:03:08.633  1000 11504 11504 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
03-24 18:03:08.633  1000 11504 11504 F DEBUG   : Build fingerprint: 'Xiaomi/bixi/bixi:15/AQ3A.241229.001/OS2.0.250322.1.VOHCNXM:user/release-keys'
03-24 18:03:08.633  1000 11504 11504 F DEBUG   : Revision: '0'
03-24 18:03:08.633  1000 11504 11504 F DEBUG   : ABI: 'arm64'
03-24 18:03:08.634  1000 11504 11504 F DEBUG   : Timestamp: 2025-03-24 18:03:07.982605034+0800
03-24 18:03:08.634  1000 11504 11504 F DEBUG   : Process uptime: 27310s
03-24 18:03:08.634  1000 11504 11504 F DEBUG   : Cmdline: com.android.systemui
03-24 18:03:08.634  1000 11504 11504 F DEBUG   : pid: 6322, tid: 6322, name: ndroid.systemui  >>> com.android.systemui <<<
03-24 18:03:08.634  1000 11504 11504 F DEBUG   : uid: 1000
03-24 18:03:08.634  1000 11504 11504 F DEBUG   : tagged_addr_ctrl: 0000000000000001 (PR_TAGGED_ADDR_ENABLE)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   : pac_enabled_keys: 000000000000000f (PR_PAC_APIAKEY, PR_PAC_APIBKEY, PR_PAC_APDAKEY, PR_PAC_APDBKEY)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   : signal 6 (SIGABRT), code -1 (SI_QUEUE), fault addr --------
03-24 18:03:08.634  1000 11504 11504 F DEBUG   : Abort message: 'FORTIFY: pthread_mutex_lock called on a destroyed mutex (0xb400007bcd377c78)'
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :     x0  0000000000000000  x1  00000000000018b2  x2  0000000000000006  x3  0000007fcf0e0be0
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :     x4  6b6b60621f6a626e  x5  6b6b60621f6a626e  x6  6b6b60621f6a626e  x7  7f7f7f7f7f7f7f7f
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :     x8  00000000000000f0  x9  0000007d430a56d8  x10 ffffff80fffffbdf  x11 0000000000000001
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :     x12 0000000067e12ddb  x13 000000007fffffff  x14 00000000007a7ab4  x15 0000007a55dfc6e8
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :     x16 0000007d4315d018  x17 0000007d431479e0  x18 0000000000003210  x19 00000000000018b2
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :     x20 00000000000018b2  x21 00000000ffffffff  x22 b400007b3d1d7ff0  x23 0000000000000007
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :     x24 0000000000000007  x25 0000007999f495a0  x26 000000007a3e61c0  x27 000000007cd40648
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :     x28 000000007db315d8  x29 0000007fcf0e0c60
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :     lr  0000007d430e27a8  sp  0000007fcf0e0bc0  pc  0000007d430e27d8  pst 0000000000001000
03-24 18:03:08.634  1000 11504 11504 F DEBUG   : 64 total frames
03-24 18:03:08.634  1000 11504 11504 F DEBUG   : backtrace:
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #00 pc 00000000000607d8  /apex/com.android.runtime/lib64/bionic/libc.so (abort+172) (BuildId: 0b77c628c9546e8a409a10c0cbfe69d9)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #01 pc 00000000000649b8  /apex/com.android.runtime/lib64/bionic/libc.so (__fortify_fatal(char const*, ...)+132) (BuildId: 0b77c628c9546e8a409a10c0cbfe69d9)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #02 pc 000000000007377c  /apex/com.android.runtime/lib64/bionic/libc.so (HandleUsingDestroyedMutex(pthread_mutex_t*, char const*)+68) (BuildId: 0b77c628c9546e8a409a10c0cbfe69d9)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #03 pc 00000000000735f4  /apex/com.android.runtime/lib64/bionic/libc.so (pthread_mutex_lock+240) (BuildId: 0b77c628c9546e8a409a10c0cbfe69d9)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #04 pc 0000000000230b1c  /system_ext/lib64/libpag.so (BuildId: 869445ec82a108c8766e16df5e1497b592bbae16)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #05 pc 00000000000e0cfc  /system_ext/lib64/libpag.so (pag::PAGLayer::duration() const+68) (BuildId: 869445ec82a108c8766e16df5e1497b592bbae16)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #06 pc 000000000010a788  /system_ext/lib64/libpag.so (Java_org_libpag_PAGLayer_duration+68) (BuildId: 869445ec82a108c8766e16df5e1497b592bbae16)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #07 pc 000000000009c3c0  /system/framework/arm64/boot.oat (art_jni_trampoline+112) (BuildId: 53b939b8d0322a13e7a958fd1e90099fd4bf985c)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #08 pc 00000000003d612c  [anon_shmem:dalvik-jit-code-cache] (offset 0x2000000) (org.libpag.PAGImageView.c+236)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #09 pc 00000000003d67f4  [anon_shmem:dalvik-jit-code-cache] (offset 0x2000000) (com.miui.clock.tiny.pets.PetsPagView.onVisibilityAggregated+292)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #10 pc 000000000074aaa8  /system/framework/arm64/boot-framework.oat (android.view.View.dispatchVisibilityAggregated+104) (BuildId: 0759674261697563e5aec77ad569de3d681fb482)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #11 pc 00000000007e606c  /system/framework/arm64/boot-framework.oat (android.view.ViewGroup.dispatchVisibilityAggregated+268) (BuildId: 0759674261697563e5aec77ad569de3d681fb482)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #12 pc 00000000007e606c  /system/framework/arm64/boot-framework.oat (android.view.ViewGroup.dispatchVisibilityAggregated+268) (BuildId: 0759674261697563e5aec77ad569de3d681fb482)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #13 pc 00000000007e606c  /system/framework/arm64/boot-framework.oat (android.view.ViewGroup.dispatchVisibilityAggregated+268) (BuildId: 0759674261697563e5aec77ad569de3d681fb482)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #14 pc 00000000007e606c  /system/framework/arm64/boot-framework.oat (android.view.ViewGroup.dispatchVisibilityAggregated+268) (BuildId: 0759674261697563e5aec77ad569de3d681fb482)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #15 pc 00000000007e12c4  /system/framework/arm64/boot-framework.oat (android.view.ViewGroup.attachViewToParent+356) (BuildId: 0759674261697563e5aec77ad569de3d681fb482)
03-24 18:03:08.634  1000 11504 11504 F DEBUG   :       #16 pc 0000000000b8cd14  [anon_shmem:dalvik-jit-code-cache] (offset 0x2000000) (androidx.recyclerview.widget.RecyclerView.access$300+100)
